src/geometry.cpp

```diff
@@ -91,11 +91,25 @@ find_cell_inner(Particle& p, const NeighborList* neighbor_list)
         break;
       }
     }
+
+    // If we're attempting a neighbor list search and fail, we
+    // now know we should return false. This will trigger an
+    // exhaustive search from neighbor_list_find_cell and make
+    // the result from that be appended to the neighbor list.
+    if (!found)
+      return false;
   }
 
   // Check successively lower coordinate levels until finding material fill
   for (;;++p.n_coord_) {
-    // If neighbor lists did not find a cell, must do exhaustive search
+    // If we did not attempt to use neighbor lists, i_cell is still C_NONE.  In
+    // that case, we should now do an exhaustive search to find the right value
+    // of i_cell.
+    //
+    // Alternatively, neighbor list searches could have succeeded, but we found
+    // that the fill of the neighbor cell was another universe. As such, in the
+    // code below this conditional, we set i_cell back to C_NONE to indicate
+    // that.
```

Comment:
One other thing I noticed when looking at the refactor of this code is that we should be resetting the `found` variable to false at the end of the `for (;;++p.n_coord_)` loop. In the recursive version, the recursive call to `find_cell_inner` resulted in a new `found` stack variable initialized to false, but in our loop version, it doesn't get reset.

Reply:
Ah yes, good observation. It makes me think we should have some regression tests that we know fail in a certain way. I think we would have gone from returning a helpful error message to a segfault as a result of this.

Reply:
Yes, it's pretty easy to demonstrate actually. If you have an invalid geometry where part of a universe is missing, `find_cell` will result in a segfault. Here's a short snippet that shows this behavior:
```Python
import openmc
import openmc.lib

m = openmc.Material()
m.add_nuclide('U235', 1.0)
m.set_density('g/cm3', 5.0)
openmc.Materials([m]).export_to_xml()

plane = openmc.XPlane()
lower_cell = openmc.Cell(fill=m, region=-plane)
# Note there's no cell filling +plane
lower_univ = openmc.Universe(cells=[lower_cell])

cyl = openmc.ZCylinder(boundary_type='reflective')
cell_outer = openmc.Cell(fill=lower_univ, region=-cyl)
geom = openmc.Geometry([cell_outer])
geom.export_to_xml()

settings = openmc.Settings()
settings.batches = 10
settings.inactive = 5
settings.particles = 1000
settings.export_to_xml()

openmc.lib.init()
openmc.lib.find_cell((0.1, 0., 0.))
```

Reply:
OK, next commit should fix this.